external/vcm/vcm/limit.py

```diff
@@ -0,0 +1,99 @@
+import xarray as xr
+from sklearn.base import BaseEstimator, TransformerMixin
+from typing import Optional, Sequence, Mapping
+
+
+class DatasetQuantileLimiter(BaseEstimator, TransformerMixin):
+    """Transformer to reduce the extremity of outliers of a dataset to quantile limits.
+    
+    Limiting is done on a variable by variable basis and along specified dimensions.
+        Limits are optionally computed on a configurable subset of the dataset to avoid
+        loading the entire dataset.
+    
+    Args:
+        alpha: two-tailed alpha for computing extrema quantiles, values beyond which
+            will be reduced to the quantile
+        limit_only: variables in the dataset to be limited; if not specified all
+            variables are limited
+
+    """
+
+    def __init__(
+        self, alpha: float, limit_only: Optional[Sequence[str]] = None,
+    ):
+        self._alpha: float = alpha
+        self._limit_only: Optional[Sequence[str]] = limit_only
+        self._upper: xr.Dataset = None
+        self._lower: xr.Dataset = None
+
+    def fit(
+        self,
+        ds: xr.Dataset,
+        feature_dims: Optional[Sequence[str]] = None,
+        fit_indexers: Optional[Mapping[str, int]] = None,
+    ) -> "DatasetQuantileLimiter":
+        """Fit the limiter on a dataset.
+        
+        Args:
+            ds: Dataset to be used to fit the limits.
+            feature_dims: Dimensions along which quantile limits should NOT be
+                computed, i.e., the resulting limits will be computed separately
+                for each coordinate along the feature_dims.
+            fit_indexer: Integer-based indexers that select a subset of `ds` on
+                which to fit limits. Indexed dimensions must not be in `feature_dims`.
+            
+        Returns: Fitted limiter
+        
+        """
+        if fit_indexers is not None and feature_dims is not None:
+            for index_dim in fit_indexers.keys():
+                if index_dim in feature_dims:
+                    raise ValueError(
+                        f"Indexer dim {index_dim} may not be in feature_dims."
+                    )
+
+        sample_ds = (ds.isel(**fit_indexers) if fit_indexers is not None else ds).load()
```

Comment:
Does this function need to know about `fit_indexers`? Seems like you should extract the subselection from this method, and call it like this:

    limiter.fit(ds.isel(fit_indexers), dims=["lon"])

this is personally clearer to me since I know what `.isel` does, but not what a fit_indexer is.

Reply:
As above the use for this is in `fit_transform` which takes a single dataset and where you want to fit on a portion of it while transforming the whole thing.  If `fit` and `transform` are always separate calls then it isn't needed. I'll remove both for simplicity then.